### PR TITLE
Abstract pagination methods and http client overriding

### PR DESCRIPTION
**Pull request by MtnWithered**
* adds abstract pagination methods - for generic usage
* allows http client overriding
* adds URI scheme argument in client constructor

### CODE REVIEW

**Comment by bhelx** [comment]
@MtnWithered I think I understand the use case for http client injection, but what would be the use cases for abstracting the pagination methods? Could you go into some more detail is your description?

File: src/main/java/com/ning/billing/recurly/RecurlyClient.java [hunk 1]
```diff
@@ -160,6 +163,13 @@ public synchronized void open() throws NoSuchAlgorithmException, KeyManagementEx
     }
 
     /**
+     * Use custom http client
+     */
+    public void open(AsyncHttpClient asyncHttpClient) {
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/RecurlyClient.java]: My only concern with this is that we expose the implementation to the public API - what if we switch client one day?

How about making `createHttpClient` protected?

**Comment by MtnWithered** [comment]
@bhelx 
I need a generic way to handle the paging in all recurly objects. It's a common feature of all these objects, so an abstract is a natural way of expressing this kind of relationship. I don't want to repeat the code for every RecurlyObjects in my client, so by using an abstraction, I can do it within a single method.
Client overriding is for reconfiguration and testing purposes.

@pierre 
Sure, it's a good idea.

**Comment by MtnWithered** [comment]
Any other suggestions here? Can we proceed with merging?

**Comment by bhelx** [comment]
@MtnWithered The code looks good to me.
@pierre any other concerns?

**Comment by pierre** [comment]
👍

**Comment by MtnWithered** [comment]
Thank you.

**Pull request merged**